File: board_game_ai/games/game.py
import numpy as np
import itertools


class Game:
    def __init__(self, *, grid=None, rows=None, cols=None):
        if grid is None:
            if rows is None or cols is None:
                raise ValueError("if grid is none, rows and cols must be provided")
            self.grid = np.zeros((rows, cols), dtype=int)
        else:
            self.grid = grid.copy()

    def state_value(self, player):
        raise NotImplementedError

    def get_valid_moves(self, player):
        return [(row, col) for row, col in itertools.product(range(self.grid.shape[0]), range(self.grid.shape[1])) if self.is_valid_move(player, (row, col))]

    def is_termal_state(self):
        raise NotImplementedError

    def play_move(self, player, move):
        raise NotImplementedError

    def is_valid_move(self, player, move):
        raise NotImplementedError

    def draw(self, screen, player):
        raise NotImplementedError

    def reset(self):
        self.grid = np.zeros_like(self.grid)

    def copy(self):
        return self.__class__(grid=self.grid)

    def mouse_pos_to_cell_num(self, posx, posy, width, height):
        return int(posy * self.grid.shape[0] / height), int(posx * self.grid.shape[1] / width)

File: js/games/game.js
export class PlayerType {
    static EMPTY = 0;
    static PLAYER = 1;
    static OPPONENT = 2;

    static otherPlayer(player) {
        return player === this.PLAYER ? this.OPPONENT : this.PLAYER;
    }
};

export class Game {
    constructor(canvas, ctx, rows, cols) {
        this.canvas = canvas;
        this.ctx = ctx;
        this.rows = rows;
        this.cols = cols;
        this.grid = [];
        for (let row = 0; row < rows; row++) {
            this.grid.push([]);
            for (let col = 0; col < cols; col++) {
                this.grid[row].push(PlayerType.EMPTY);
            }
        }
        this.maxDepth = 1e8;
        this.type = Game;
    }
    simulateMove(player, move) {
        const nextGame = this.copy();
        nextGame.playMove(player, move);
        return nextGame;
    }
    copy() {
        const grid = [];
        for (let row = 0; row < this.rows; row++) {
            grid.push([]);
            for (let col = 0; col < this.cols; col++) {
                grid[row].push(this.grid[row][col]);
            }
        }
        // const nextGame = Object.create(this)
        // const nextGame = new Tictactoe(grid);
        const nextGame = new this.type(this.canvas, this.ctx, grid);
        return nextGame;
    }
    playMove(player, move) { }
    isValidMove(player, move) { }
    getValidMoves(player) { }
    evaluate(player) { }
    mouseToMove(x, y) { }
    draw(winner, player) { }
}